src/nix/path-info.cc

```diff
@@ -9,6 +9,75 @@
 #include <nlohmann/json.hpp>
 
 using namespace nix;
+using nlohmann::json;
+
+/**
+ * @return the total size of a set of store objects (specified by path),
+ * that is, the sum of the size of the NAR serialisation of each object
+ * in the set.
+ */
+static uint64_t getStoreObjectsTotalSize(Store & store, const StorePathSet & closure)
+{
+    uint64_t totalNarSize = 0;
+    for (auto & p : closure) {
+        totalNarSize += store.queryPathInfo(p)->narSize;
+    }
+    return totalNarSize;
+}
+
+
+/**
+ * Write a JSON representation of store object metadata, such as the
+ * hash and the references.
+ *
+ * @param showClosureSize If true, the closure size of each path is
+ * included.
+ */
+static json pathInfoToJSON(
+    Store & store,
+    const StorePathSet & storePaths,
+    bool showClosureSize)
+{
+    json::object_t jsonAllObjects = json::object();
+
+    for (auto & storePath : storePaths) {
+        json jsonObject;
+
+        try {
+            auto info = store.queryPathInfo(storePath);
+
+            jsonObject = info->toJSON(store, true, HashFormat::SRI);
+
+            if (showClosureSize) {
+                StorePathSet closure;
+                store.computeFSClosure(storePath, closure, false, false);
+
+                jsonObject["closureSize"] = getStoreObjectsTotalSize(store, closure);
+
+                if (auto * narInfo = dynamic_cast<const NarInfo *>(&*info)) {
```

Comment:
```
src/nix/path-info.cc:57:28: warning: unused variable ‘narInfo’ [-Wunused-variable]
   57 |                 if (auto * narInfo = dynamic_cast<const NarInfo *>(&*info)) {
      |                            ^~~~~~~
```

Was something lost along the way, or is `narInfo` truly redundant? Or is the whole conditional redundant?

Reply:
`computeFSClosure` is transitive-reflexive I think so we do not need `narInfo` (`auto & p : closure` below will get us what we need).

However we do still need the conditional because the invariant is "the transitive references are only expected to have `NarInfo` if the root has `NarInfo`".

Reply:
(The original code would just skip store objects without a `NarInfo`, making it possible to get a bogus `closureDownloadSize` silently if an internal invariant was broken.)